solr/core/src/java/org/apache/solr/search/grouping/distributed/responseprocessor/SearchGroupShardResponseProcessor.java

```diff
@@ -108,9 +114,11 @@ public void process(ResponseBuilder rb, ShardRequest shardRequest) {
             .put(SolrQueryResponse.RESPONSE_HEADER_PARTIAL_RESULTS_KEY, Boolean.TRUE);
         continue; // continue if there was an error and we're tolerant.
       }
-      maxElapsedTime = (int) Math.max(maxElapsedTime, srsp.getSolrResponse().getElapsedTime());
-      NamedList<NamedList<?>> firstPhaseResult =
-          (NamedList<NamedList<?>>) srsp.getSolrResponse().getResponse().get("firstPhase");
+      maxElapsedTime = (int) Math.max(maxElapsedTime, solrResponse.getElapsedTime());
+      NamedList<NamedList<?>> firstPhaseResult = getFirstPhaseFromShardResponse(rb, srsp);
+      if (firstPhaseResult == null) {
```

Comment:
noting that `secondPhase` is already checking for null -- https://github.com/apache/solr/blob/releases/solr/9.0.0/solr/core/src/java/org/apache/solr/search/grouping/distributed/responseprocessor/TopGroupsShardResponseProcessor.java#L126 -- though it appears to not consider the partial response flag

Reply:
Thanks added util class to help with these. I addressed a bunch of other places that follow the same pattern.